src/core/pyproxy.c

```diff
@@ -696,6 +697,16 @@ _pyproxy_apply(PyObject* callable,
   return idresult;
 }
 
+bool
+_iscoroutinefunction(PyObject* f)
+{
+  PyObject* result = PyObject_CallOneArg(iscoroutinefunction, f);
+  if (!result) {
+    PyErr_Clear();
+  }
+  return Py_IsTrue(result);
+}
+
```

Comment:
so, if I understand correctly, now we are calling the app-level `inspect.iscoroutinefunction()` every time we call a PyProxy, is it correct?
I don't know what is the actual performance penalty, but it might be non-negligible.

`inspect.iscoroutinefunction` seems very trivial, although `_has_code_flag()` is not completely trivial:
https://github.com/antocuni/cpython/blob/deaf509e8fc6e0363bd6f26d52ad42f976ec42f2/Lib/inspect.py#L391-L414

maybe it is worth considering adding a fast path for common cases like "plan function" and "plain bound method", and check the flags directly in C. Something along this line (mostly pseudocode):
```c
bool iscoroutinefunciton(PyObject *obj) {
    if (PyFunction_Check(obj)) {
        PyFunctionObject *func = (PyFunctionObject*)obj;
        return func->func_code->co_flags & CO_COROUTINE;
    }
    else if (PyMethod_Check(obj)) {
        ...
    }
    else {
        // fallback to call inspect.iscoroutinefunction as we do now
    }
}
```

Reply:
Looks like this Python 3.12 added [`markcoroutinefunction`](https://docs.python.org/3.12/library/inspect.html#inspect.markcoroutinefunction) which adds a possibility to get false negatives. But we can check if `hasattr(o, "_is_coroutine_marker")` and fall back to `inspect.iscoroutinefunction` in that case.

This does add a bit of extra maintenance burden when updating Python versions...

Reply:
it's debatable whether they are false negatives.
I'll write my thoughts in the main discussion